common/src/main/java/io/pravega/common/util/AsyncIterator.java

```diff
@@ -0,0 +1,55 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.util;
+import io.pravega.common.concurrent.Futures;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+/**
+ * Defines an Iterator for which every invocation results in an async call with a delayed response.
+ *
+ * @param <T> Element type.
+ */
+public interface AsyncIterator<T> {
```

Comment:
I think the assumption here is that order does not matter. In the case it does matter, I believe we can't guarantee that the application will process the futures in order. Could you clarify on the primary reason for having this asynchronous iterator?

Reply:
So you can make IO calls upon calling `getNext` and give the caller the option on whether to block while waiting (and hogging a thread) or not. Using the classic `Iterator` you do not have this option and it would have had to block on the IO call for you.